### ISSUE TITLE
storage pools - a cautionary tale

### ISSUE THREAD

**Issue by danielrobbins** [lxc/lxd#6057]
I experienced some upgrade woes when I moved to lxd-3.14, related to ZFS storage pools.

With other storage pools, at least of the btrfs variety, if the pool is not available then the containers that use the storage pool fail to start, but all other containers start.

In this particular scenario, I had upgraded a server to a new debian-sources-lts kernel release, and I had decided to not rebuild the ZFS modules as I am not officially using it anymore. Upon reboot, lxd would not start. I do not have the exact message, but it appears by design that it would not start since it could not initialize ZFS which was required by one stray storage pool.

Yes, it appears that I had a 'stray' ZFS storage pool that had one unused image in it that was left over. Because it was still part of the lxd configuration, lxd tried to initialize ZFS, failed, and then simply aborted and would refuse to run. So I had approx. 50 containers that would not start. And I couldn't use the "lxc" commands to fix my configuration since lxd wouldn't start! :(

I attempted to work around this issue by manually tweaking the sqlite db to remove the ZFS storage pool. Documentation on this appears to be non-existent, and my attempts failed anyway -- the storage pool kept re-appearing in the dqlite db despite my best attempts to zap it.

Finally I succumbed to my need to simply get the containers up and running, and had to cave to lxd's needs -- I installed zfs and friends so that zfs would be available. I thought that this would make lxd happy.

Unfortunately, even with ZFS available (kernel module and tools), ZFS now complained that my long-gone ZFS storage pool didn't exist -- and then would abort and not start! So I still couldn't start any containers nor use the "lxc" commands to fix my configuration.

Desperate, I worked around this issue by creating a 300MB loopback zvol with the name expected by lxd, and then lxd happily started and I was able to start all my containers.

I am sharing this cautionary tale so you can understand that the current behavior may be less-than-ideal in production environments, particularly when someone is migrating away from ZFS.

And yes, in theory my ZFS-based storage pool shouldn't have existed in my lxd configuration -- and I am not sure how it continued to exist -- apparently it's possible to zap a zvol underneath lxd and as long as no containers are using it, lxd is none the wiser until it has to (re)start and realizes that zvol no longer exists.

What I would have preferred -- if a zvol-based storage volume is not available, allow lxd to start and attempt to start all containers it can anyway. In my case, this would have allowed ALL my containers to start as none needed this zfs storage volume in the first place. Then, print out some kind of loud warning to the console or logs so that I know that lxd is running in a degraded state and was unable to initialize at least one storage volume.

(Side-note: when a storage volume is unavailable and containers do not start (ie. with btrfs storage volumes where lxd seems to still start) the logging related to the storage volume not being available appears to be non-existent -- you sort of have to derive the actual error condition from manual troubleshooting rather than getting a clear "underlying storage volume not available" in the logs. Ideally this would print to the console even if you don't have debugging enabled. If a storage volume is not available, ideally, even a basic "lxc start container" would print a clear "underlying storage volume does not appear to be available" to assist troubleshooting.)

**Comment by stgraber** [comment]
Normally LXD will happily start with missing storage, the 3.15 upgrade however wasn't such a case as one time migration logic had to be performed on all storage pools prior to the daemon being able to safely operate.

LXD's database file is generated from the raft log entries and so cannot be user modified (it just gets reset on startup), however we have a documented process to run commands immediately as the database is brought up, this would have let you run the suitable `DELETE FROM storage_pools;` to unblock this situations: https://lxd.readthedocs.io/en/latest/database/#running-custom-queries-at-lxd-daemon-startup

**Comment by stgraber** [comment]
Sending a branch to have ZFS properly wrap its errors:

Before:
```
stgraber@castiana:~$ lxc start test
Error: Common start logic: no such file or directory
Try `lxc info --show-log test` for more info
```

After:
```
stgraber@castiana:~$ lxc start test
Error: Common start logic: Failed to mount ZFS dataset "castiana/lxd/containers/test" onto "/var/lib/lxd/storage-pools/default/containers/test": no such file or directory
Try `lxc info --show-log test` for more info
```

**Comment by danielrobbins** [comment]
Thanks for the reply. Oh, lxd sql looks like what I needed. 

I think the ideal way to handle the one-time upgrade scenario would be to print a clear error message and then say something like "Storage pool X not found. Cannot perform one-time update steps for lxd x.y. Lxd will exit. Please correct the issue by making the storage pool available or use the 'lxc sql' command to remove storage pool if no longer needed."

I think it's really good practice to provide verbose and clear error messages, particularly when lxd is doing something special such as an upgrade and an administrator such as myself may be confused by lxd's unusual behavior.
